numpy/core/src/multiarray/nditer_api.c

```diff
@@ -2529,15 +2532,6 @@ npyiter_copy_to_buffers(NpyIter *iter, char **prev_dataptrs)
                 skip_transfer = 1;
             }
 
-            /* If the data type requires zero-inititialization */
-            if (PyDataType_FLAGCHK(dtypes[iop], NPY_NEEDS_INIT)) {
-                NPY_IT_DBG_PRINT("Iterator: Buffer requires init, "
-                                    "memsetting to 0\n");
-                memset(ptrs[iop], 0, dtypes[iop]->elsize*op_transfersize);
-                /* Can't skip the transfer in this case */
-                skip_transfer = 0;
```

Comment:
Should the code to reset `skip_transfer = 0` remain? Is there ever a case where both `reuse_reduce_loops && prev_dataptrs[iop] == ad_ptrs[iop]` and `PyDataType_FLAGCHK(dtypes[iop], NPY_NEEDS_INIT)`

Reply:
The case you are saying seems very plausible to me. Note that `NPY_NEEDS_INIT` is pretty much equivalent to "object dtype", so it isn't much of a limitation.

Since this is limited to object-dtype and for all other dtypes the transfer is skipped, I would be surprised if it was necessary.  But, I have to dig deeper :/...   In the case of an object read-write loop, we have to be sure that the buffer is not cleared to be able to reuse it (this translates to "move references" not being used to write it back to the array).  I guess I have to try and make sure we have a test for this (we may well have, but object dtype tests tend to be a bit spotty).

Reply:
Well, I think you are right, we have to copy always if:
```
                    transferinfo[iop].write.func != NULL &&
                    PyDataType_REFCHK(transferinfo[iop].write.descriptors[0])
```
(or similar).  But I am still struggling with a reproducer, maybe I have to set up a weirdly complicated nditer manually...

Reply:
I guess its also a bit confusing that `NEEDS_INIT` and `HASREF` are really synonymous...  Anything that has reference must require initialization.  But I can't see of a reason why the other way around can make sense (requiring initialization but for a reason other than `HASREF`).

Reply:
The easiest thing to do here would be to remove only line 2536, maybe with a `/* TODO: do we really need this? */`

Reply:
Yeah, that was just the, not really ideal, try to rephrase the current logic into something slightly more to the point.  I prefer to rephrase it (although by just making it `if (!skip_transfer || PyDataType_REFCHK(dtypes[iop])) {`)

I finally found a reproducer for at least one way this can fail with reductions...